fawltydeps/packages.py

```diff
@@ -310,20 +310,45 @@ class TemporaryPipInstallResolver(BasePackageResolver):
 
     @staticmethod
     @contextmanager
-    def temp_installed_requirements(requirements: List[str]) -> Iterator[Path]:
+    def temp_installed_requirements(
+        requirements: List[str],
+        ignore_errors: bool = False,
+        retry_each: bool = False,
+    ) -> Iterator[Path]:
         """Create a temporary venv and install the given requirements into it.
 
         Provide a path to the temporary venv into the caller's context in which
         the given requirements have been `pip install`ed. Automatically remove
         the venv at the end of the context.
+
+        By default, if `pip install` fails, we will propagate the resulting
+        CalledProcessError from here. If 'ignore_errors' is enabled we instead
+        log a warning and continue into the caller's context. It is then up to
+        the caller to handle the missing requirements in the temporary venv.
+
+        Furthermore, if 'retry_each' is enabled, and if the `pip install` fails,
+        we will try to `pip install` each given requirement in order (logging
+        any errors that occur) before we finally enter the caller's context.
+        This should give the caller access to as many as possible of the given
+        requirements.
         """
         with tempfile.TemporaryDirectory() as tmpdir:
             venv_dir = Path(tmpdir)
             venv.create(venv_dir, clear=True, with_pip=True)
-            subprocess.run(
-                [f"{venv_dir}/bin/pip", "install", "--no-deps"] + requirements,
-                check=True,  # fail if any of the commands fail
-            )
+            argv = [f"{venv_dir}/bin/pip", "install", "--no-deps"] + requirements
+            proc = subprocess.run(argv, check=not ignore_errors)
```

Comment:
I do not understand why a `CalledProcessError` is not raised, could you explain it? According to documentation, https://docs.python.org/3/library/subprocess.html#using-the-subprocess-module:
 > If check is true, and the process exits with a non-zero exit code, a CalledProcessError exception will be raised. Attributes of that exception hold the arguments, the exit code, and stdout and stderr if they were captured.

Reply:
Oh, I know why! This case is never tested in tests, as `ignore_errors` is set to `True` :D

In that case we can either drop `ignore_errors` parameter and have:
```python
            proc = subprocess.run(argv, check=False)
```
Or test this case properly in tests, because it will result in some more guardian code around `CalledProcessError`.

I would opt for option 1, as I think that what we do here is an equivalent of handling this error.

Reply:
Yes, I wrote the patch to be minimally invasive in terms of preserving current behavior unless the extra flags are passed. But since there is only one caller, and it always passes both flags as `True`, we can drop the flags, and hardcode the new behavior instead. Will fix.